Improve TFTP detection with options

thomasjwinter opened the pull request:
TFTP read/write requests can have optional strings after the filename and mode such as block size and transfer size to speed the transfer. There is also the Option Acknowledgement opcode that contains the option strings that were accepted.

nDPI fails to recognize tftp transactions using options as tftp.

thomasjwinter committed:
tftp: rework request checking to account for options

Read and write requests were not handling options.
The existing code inspecting request messages assumed that the string
before the end of the payload was the mode and that the filename
length depended on the mode length.
However the first two strings are the filename and mode which can be
followed by any number of option strings.

Rework the checking of the filename and mode to just search the first
two strings which should always exist. Any options after that are
ignored.

Absence of the filename or mode is now excludes the TFTP protocol.
Absence of the filename no longer is considered malformed packet
because it can match other protocols falsely.

thomasjwinter committed:
tftp: check for Option Acknowledgements

TFTP Option Acknowledgement packets were being excluded.
When a read/write request contains options, an Option Acknowledgement
is returned that contains the option strings that the transaction
will use.
The options sent in the request are not compared with what was
acknowledged.

utoni commented on src/lib/protocols/tftp.c:
```diff
@@ -65,24 +65,55 @@ static void ndpi_search_tftp(struct ndpi_detection_module_struct *ndpi_struct,
-          for(mode_idx = 0; mode_idx < NDPI_ARRAY_LENGTH(possible_modes); ++mode_idx)
+          int i;
+          size_t filename_len = 0;
+          char *string;
```
```suggestion
          const char *string;
```

utoni commented:
Please take a look at the CI: https://github.com/ntop/nDPI/actions/runs/5933443642/job/16095465366?pr=2074

Seems we are loosing one TFTP flow detection.

thomasjwinter commented:
It looks like the tftp.pcap was changed in commit 1a01e8dc6 where it added two invalid read requests that don't abide by the rfc. You can open it in wireshark and see that it interprets them strangely. The second read request has 11 zero bytes following the opcode then the mode which is interpreted as an option. I believe this should have been interpreted as a malformed packet with zero length filename so this patch should be more correct.

utoni commented:
Ok, please update the test results with `NDPI_FORCE_UPDATING_UTESTS_RESULTS=1 NDPI_DISABLE_FUZZY=1 ./tests/do.sh` and upload the changes.

thomasjwinter commented:
Rerunning the tests and inspecting closer I see the cause.
My new code was excluding the flow as no mode was found due the new NULL terminated string checking. Previous code was only looking at the end of the payload to get the mode then doing the filename risk check.

Attempting to move the filename risk check above the mode_found exclude check results in the stun_classic.pcap failing as it happens to have the 0x0001 opcode bytes in the right place but the following bytes aren't printable characters.

What do you suggest I should do?
Could remove the malformed packet risk and exclude tftp if the filename or mode is not found.
Should we be excluding if not standard TFTP UDP port 69?

utoni commented:
I think you can safely remove the malformed packet risk. I remember setting this but the TFTP detection patterns are rather sparse. So I guess it is better to have a "stable" detection instead of a risk set.

IvanNardi commented:
> I think you can safely remove the malformed packet risk. I remember setting this but the TFTP detection patterns are rather sparse. So I guess it is better to have a "stable" detection instead of a risk set.

+1

thomasjwinter committed:
tftp: update pcap results

The two malformed TFTP packets are no longer considered as risk
and instead match by port only.
This is because the TFTP detection was rather sparse so could
match on several other protocols if the first two opcode bytes
happened to match.

thomasjwinter committed:
tftp: check incrementation for DATA and ACK packets

The 2 bytes following the opcode for DATA and ACK packets are the
block number and this should be incrementing every packet.
We should check to see that this is occurring otherwise false matches
can occur, eg L2TPv3 over UDP matches the DATA opcode but the next two
bytes are always zero.

Remove the DATA max block size assumption since this can be false if
the blksize option is used to increase it.

Fixes #2070

utoni commented on tests/cfgs/default/result/tftp.pcap.out:
```diff
@@ -25,7 +26,7 @@ TFTP	107	31296	7
 	2	UDP 172.28.5.170:62058 <-> 172.28.5.91:44618 [proto: 96/TFTP][IP: 0/Unknown][ClearText][Confidence: DPI][DPI packets: 4][cat: DataTransfer/4][2 pkts/92 bytes <-> 2 pkts/1116 bytes][Goodput ratio: 9/92][0.00 sec][Risk: ** Known Proto on Non Std Port **][Risk Score: 50][PLAIN TEXT (BCCCCCC)][Plen Bins: 50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, ...
 	3	UDP 192.168.0.253:50618 -> 192.168.0.10:69 [proto: 96/TFTP][IP: 0/Unknown][ClearText][Confidence: DPI][DPI packets: 1][cat: DataTransfer/4][1 pkts/62 bytes -> 0 pkts/0 bytes][Goodput ratio: 32/0][< 1 sec][Filename: rfc1350.txt][Risk: ** Unidirectional Traffic **][Risk Score: 10][Risk Info: No server to client traffic][PLAIN TEXT (1350.txt)][Plen Bins: 100,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, ...
-	4	UDP 172.28.4.53:54626 -> 172.16.5.170:69 [proto: 96/TFTP][IP: 0/Unknown][ClearText][Confidence: DPI][DPI packets: 1][cat: DataTransfer/4][1 pkts/61 bytes -> 0 pkts/0 bytes][Goodput ratio: 31/0][< 1 sec][Risk: ** Malformed Packet **** Unidirectional Traffic **][Risk Score: 20][Risk Info: No server to client traffic / Invalid TFTP RR/WR header: Source/Destination file missing][Plen Bins: 100,0,0 ...
+	4	UDP 172.28.4.53:54626 -> 172.16.5.170:69 [proto: 96/TFTP][IP: 0/Unknown][ClearText][Confidence: Match by port][DPI packets: 1][cat: DataTransfer/4][1 pkts/61 bytes -> 0 pkts/0 bytes][Goodput ratio: 31/0][< 1 sec][Risk: ** Unidirectional Traffic **][Risk Score: 10][Risk Info: No server to client traffic][Plen Bins: 100,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, ...
```
Shouldn't that be `[Confidence: DPI]` ?

utoni commented on tests/cfgs/default/result/tftp.pcap.out:
```diff
@@ -25,7 +26,7 @@ TFTP	107	31296	7
 	5	UDP 172.28.4.53:54627 -> 172.16.5.170:69 [proto: 96/TFTP][IP: 0/Unknown][ClearText][Confidence: DPI][DPI packets: 1][cat: DataTransfer/4][1 pkts/61 bytes -> 0 pkts/0 bytes][Goodput ratio: 31/0][< 1 sec][Filename: sysman.lis][Risk: ** Unidirectional Traffic **][Risk Score: 10][Risk Info: No server to client traffic][PLAIN TEXT (sysman.lis)][Plen Bins: 100,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 ...
 	6	UDP 172.28.5.91:44618 -> 172.28.5.170:69 [proto: 96/TFTP][IP: 0/Unknown][ClearText][Confidence: DPI][DPI packets: 1][cat: DataTransfer/4][1 pkts/60 bytes -> 0 pkts/0 bytes][Goodput ratio: 30/0][< 1 sec][Filename: zz.bin][Risk: ** Unidirectional Traffic **][Risk Score: 10][Risk Info: No server to client traffic][PLAIN TEXT (zz.bin)][Plen Bins: 100,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 ...
-	7	UDP 172.28.4.53:54632 -> 172.16.5.170:69 [proto: 96/TFTP][IP: 0/Unknown][ClearText][Confidence: DPI][DPI packets: 1][cat: DataTransfer/4][1 pkts/51 bytes -> 0 pkts/0 bytes][Goodput ratio: 17/0][< 1 sec][Risk: ** Malformed Packet **** Unidirectional Traffic **][Risk Score: 20][Risk Info: No server to client traffic / Invalid TFTP RR/WR header: Source/Destination file missing][Plen Bins: 100,0,0 ...
+	7	UDP 172.28.4.53:54632 -> 172.16.5.170:69 [proto: 96/TFTP][IP: 0/Unknown][ClearText][Confidence: Match by port][DPI packets: 1][cat: DataTransfer/4][1 pkts/51 bytes -> 0 pkts/0 bytes][Goodput ratio: 17/0][< 1 sec][Risk: ** Unidirectional Traffic **][Risk Score: 10][Risk Info: No server to client traffic][Plen Bins: 100,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, ...
```
same here

thomasjwinter commented:
As discussed previously, those are the 2 packets with incorrect formats so are now being excluded.

IvanNardi commented on src/include/ndpi_typedefs.h:
```diff
@@ -1551,6 +1551,8 @@ struct ndpi_flow_struct {
     struct {
       char filename[128];
+      u_int16_t data_num;
+      u_int16_t ack_num;
```
This is wrong. These two variables need to be defined in `struct ndpi_flow_udp_struct` (like `tftp_stage`).
Reason: since `ndpi_flow_struct->protos` is an union, we can't write to it until we have a final classification, i.e. until we call `ndpi_int_tftp_add_connection()`.

Could you rebase and update the code, please?

IvanNardi reviewed (changes requested):
Thanks for the contribution. Could you try to fix the code, as reported below, please?

IvanNardi commented:
Thanks for the fixes!
